src/particles/plasma/PlasmaParticleContainer.cpp

```diff
@@ -624,30 +626,85 @@ LaserIonization (const int islice,
         amrex::Gpu::DeviceScalar<uint32_t> ip_elec(0);
         uint32_t * AMREX_RESTRICT p_ip_elec = ip_elec.dataPtr();
 
-        // This kernel adds the new ionized electrons to the Plasma Particle Container
-        amrex::ParallelFor(num_ions,
-            [=] AMREX_GPU_DEVICE (long ip) {
+        // This kernel supports multiple deposition orders (0, 1, 2, 3) at compile time
+        // and calculates the momentum of the ionized electron based on equations (B8) and (B9)
+        // from the Massimo, 2020 article. It computes the energy of the emitted electron
+        // and assigns the resulting properties (momentum, position, etc.) to the new electrons
+        // created in the plasma container.
+        amrex::AnyCTO(
+            amrex::TypeList<
+                amrex::CompileTimeOptions<0, 1, 2, 3>
+            >{}, {
+                Hipace::m_depos_order_xy
+            },
+            [&] (auto cto_func) {
+                amrex::ParallelForRNG(num_ions, cto_func);
+            },
+            [=] AMREX_GPU_DEVICE (long ip, const amrex::RandomEngine& engine,
+                                  auto depos_order_xy) {
 
             if(p_ion_mask[ip] != 0) {
+
+                // Avoid temp slice
+                const amrex::Real xp = x_prev[ip];
+                const amrex::Real yp = y_prev[ip];
+
+                if (amrex::ConstParticleIDWrapper(idcpup[ip]) < 0 ||
+                    !laser_bounds.contains(xp, yp)) return;
+
+                Complex A = 0;
+                Complex A_dx = 0;
+                Complex A_dzeta = 0;
+
+                doLaserGatherShapeN<depos_order_xy>(xp, yp, A, A_dx, A_dzeta, laser_arr,
+                    dx_inv, dy_inv, dzeta_inv, x_pos_offset, y_pos_offset);
+
+                const Complex Et = I * A * omega0 + A_dzeta * phys_const.c; // transverse component
+                const Complex El = - A_dx * phys_const.c; // longitudinal component
+
+                amrex::Real Ep = std::sqrt( amrex::abs(Et*Et) + amrex::abs(El*El) );
+                Ep *= phys_const.m_e * phys_const.c / phys_const.q_e;
+                Ep *= E0;
+
+                amrex::Real ux;
+                amrex::Real uy;
+                amrex::Real uz;
+                const int ion_lev_loc = ion_lev[ip];
+
+                if (linear_polarization) {
+                    amrex::Real delta = std::sqrt(Ep) * laser_dp_prefactor[ion_lev_loc-1];
+                    amrex::Real delta2 = delta * delta;
+                    amrex::Real width_p = amrex::abs(A) * delta * (1 - (3./4.) * delta2 - (3./2.) * delta2 + laser_dp_second_prefactor[ion_lev_loc-1] * delta2); // equation (14) from C. Schroeder art.
```

Comment:
`_rt` avoids some type conversions, we can discuss offline.
```suggestion
                    // equation (14) from Schroeder PRAB 2014
                    amrex::Real const width_p = amrex::abs(A) * delta * (1._rt - (3._rt/4._rt) * delta2 - (3._rt/2._rt) * delta2 + laser_dp_second_prefactor[ion_lev_loc-1] * delta2);
```